File: src/libs/m3/session/Session.cc
/*
 * Copyright (C) 2015, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <m3/session/Session.h>
#include <m3/Syscalls.h>
#include <m3/VPE.h>

namespace m3 {

void Session::connect(const String &service, xfer_t arg, capsel_t cap) {
    if(cap == ObjCap::INVALID)
        cap = VPE::self().alloc_cap();
    Errors::Code res = Syscalls::get().createsess(cap, service, arg);
    if(res == Errors::NONE)
        sel(cap);
}

void Session::delegate_for(VPE &vpe, const KIF::CapRngDesc &crd, KIF::ExchangeArgs *args) {
    Syscalls::get().delegate(vpe.sel(), sel(), crd, args);
}

void Session::obtain_for(VPE &vpe, const KIF::CapRngDesc &crd, KIF::ExchangeArgs *args) {
    vpe.mark_caps_allocated(crd.start(), crd.count());
    Syscalls::get().obtain(vpe.sel(), sel(), crd, args);
}

}

File: src/libs/rust/src/session/session.rs
use cap::{Capability, Flags, Selector};
use errors::Error;
use kif::{CapRngDesc, cap};
use syscalls;
use vpe;

pub struct Session {
    cap: Capability,
}

impl Session {
    pub fn new(name: &str, arg: u64) -> Result<Self, Error> {
        let sel = vpe::VPE::cur().alloc_cap();
        try!(syscalls::create_sess(sel, name, arg));
        Ok(Session {
            cap: Capability::new(sel, Flags::empty()),
        })
    }

    pub fn new_bind(sel: Selector) -> Self {
        Session {
            cap: Capability::new(sel, Flags::KEEP_CAP),
        }
    }

    pub fn sel(&self) -> Selector {
        self.cap.sel()
    }

    pub fn delegate_obj(&self, sel: Selector) -> Result<(), Error> {
        let crd = CapRngDesc::new_from(cap::Type::OBJECT, sel, 1);
        let sargs = [];
        let mut rargs = [];
        self.delegate(crd, &sargs, &mut rargs).map(|_| ())
    }

    pub fn delegate(&self, crd: CapRngDesc, sargs: &[u64], rargs: &mut [u64]) -> Result<usize, Error> {
        syscalls::delegate(self.sel(), crd, sargs, rargs)
    }

    pub fn obtain(&self, count: u32, sargs: &[u64], rargs: &mut [u64]) -> Result<(usize, CapRngDesc), Error> {
        let caps = vpe::VPE::cur().alloc_caps(count);
        let crd = CapRngDesc::new_from(cap::Type::OBJECT, caps, count);
        let num = try!(syscalls::obtain(self.sel(), crd, sargs, rargs));
        Ok((num, crd))
    }
}
